pymc/step_methods/state.py

```diff
@@ -0,0 +1,98 @@
+#   Copyright 2024 The PyMC Developers
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+from copy import deepcopy
+from dataclasses import dataclass, fields
+from typing import Any
+
+import numpy as np
+
+
+class MetaDataClassState(type):
+    def __new__(cls, name, bases, namespace):
+        return dataclass(eq=False, kw_only=True)(super().__new__(cls, name, bases, namespace))
+
+
+class DataClassState(metaclass=MetaDataClassState):
+    def __eq__(self, other: Any) -> bool:
+        if self.__class__ != other.__class__:
+            return False
+        this_fields = set([f.name for f in fields(self)])
+        other_fields = set([f.name for f in fields(other)])
+        for field in this_fields:
+            self_val = getattr(self, field)
+            other_val = getattr(other, field)
+            if not DataClassState.compare_values(self_val, other_val):
+                return False
+        return this_fields == other_fields
+
+    @staticmethod
+    def compare_values(v1: Any, v2: Any) -> bool:
+        if v1.__class__ != v2.__class__:
+            return False
+        if isinstance(v1, (list, tuple)):  # noqa: UP038
+            return len(v1) == len(v2) and all(
+                DataClassState.compare_values(v1i, v2i) for v1i, v2i in zip(v1, v2)
+            )
+        elif isinstance(v1, dict):
+            if set(v1) != set(v2):
+                return False
+            return all(DataClassState.compare_values(v1[k], v2[k]) for k in v1)
+        elif isinstance(v1, np.ndarray):
+            return bool(np.array_equal(v1, v2, equal_nan=True))
+        elif isinstance(v1, np.random.Generator):
+            return DataClassState.compare_values(v1.bit_generator.state, v2.bit_generator.state)
+        else:
+            return v1 == v2
+
+
+class WithSamplingState:
+    _state_class: type[DataClassState] = DataClassState
+
+    @property
+    def sampling_state(self) -> DataClassState:
+        state_class = self._state_class
+        kwargs = {}
+        for field in fields(state_class):
+            val = getattr(self, field.name)
+            if isinstance(val, WithSamplingState):
+                _val = val.sampling_state
+            else:
+                _val = val
+            kwargs[field.name] = deepcopy(_val)
+        return state_class(**kwargs)
+
+    @sampling_state.setter
```

Comment:
At the moment, it's only used in the tests. Calling `step_method.sampling_state = state` will set the step method's attributes to the proper values represented by the provided `state`. The goal is that we'll use it when we want to jump start the sampler to some past sampling state. The workflow I have in mind is this:
1. Have the model already built
2. Have the samplers already built
3. If someone provides a past sampling state from which to start, set the sampler's state to that
4. Start sampling using the sampler's set state and collect the results

Reply:
I prefer an explicit method for that functionality? Something like `step_method.set_state(state)`?

It's 100% subjective opinion though

Reply:
I agree to disagree on this one. The `set_state` method looks like any other method call and it’s not explicitly saying that it won’t have any return value. The attribute assignment syntax on the other hand is much more explicit on its intent.

Reply:
So why is there a set_rng? The same argument you used against set_state would apply? Also, in my experience properties have always be a PITA because at some point we figure out we would like kwargs and there's no way to refactor a property into a method with back_compat.

Reply:
I added a `set_rng` method because the subclasses are supposed to overload it. The HMC step methods have their own `rng` and their `potential` also have a spawned `rng`. Setting the `rng` had to work differently than with the rest of the step methods, so I decided to make it a method instead of a `property`. Anyway, I still prefer the `sampling_state` as a property mixin. If we eventually realize that we don't want to, we can always add a new `set_state` method and issue a `DeprecationWarning` or `RuntimeError` in the property setter.

Reply:
Let me try one last argument. Setting a property (which looks like an attribute) does not make me intuitively think that it will actually affect the sampler. Special in this case where the attribute (sampler.sample_sate) is actually a read-only of the internal state, not the state itself. 

It's like if you have a PyMC model, which has the attribute `model.datalogp`. I wouldn't expect  `model.datalogp = 0` to be a valid way of overriding the model logp. Of course if I read the source code I'll be able to figure out, but from just reading use code I wouldn't think it would do what I want it to do.